cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/UnreachableEOGPass.kt

```diff
@@ -28,71 +28,196 @@ package de.fraunhofer.aisec.cpg.passes
+ * it is always true, then the edge to the code after the loop receives the [state]
+ * [Reachability.UNREACHABLE]. If the condition is always false, then the edge to the loop body
+ * receives the [state] [Reachability.UNREACHABLE]. All other cases simply copy the state which led
+ * us here.
+ */
+private fun handleWhileStatement(
+    enteringEdge: PropertyEdge<Node>,
+    n: WhileStatement,
+    state: State<PropertyEdge<Node>, Reachability>
+) {
+    /*
+     * Note: It does not understand that code like
+     * x = true; while(x) {...; x = false;}
+     * makes the loop execute at least once.
+     * Apparently, the CPG does not offer the required functionality to
+     * differentiate between the first and subsequent evaluations of the
+     * condition.
+     */
+    val evalResult = ValueEvaluator().evaluate(n.condition)
+
+    val (unreachableEdge, remainingEdges) =
         if (evalResult is Boolean && evalResult == true) {
-            n.nextEOGEdges
-                .firstOrNull { e -> e.getProperty(Properties.INDEX) == 1 }
-                ?.addProperty(Properties.UNREACHABLE, true)
+            Pair(
+                n.nextEOGEdges.firstOrNull { e -> e.getProperty(Properties.INDEX) == 1 },
+                n.nextEOGEdges.filter { e -> e.getProperty(Properties.INDEX) != 1 }
+            )
         } else if (evalResult is Boolean && evalResult == false) {
-            n.nextEOGEdges
-                .firstOrNull { e -> e.getProperty(Properties.INDEX) == 0 }
-                ?.addProperty(Properties.UNREACHABLE, true)
+            Pair(
+                n.nextEOGEdges.firstOrNull { e -> e.getProperty(Properties.INDEX) == 0 },
+                n.nextEOGEdges.filter { e -> e.getProperty(Properties.INDEX) != 0 }
+            )
+        } else {
+            Pair(null, n.nextEOGEdges)
         }
-    }
 
-    override fun cleanup() {
-        // nothing to do
+    if (unreachableEdge != null) {
+        // This edge is definitely unreachable
+        state.push(unreachableEdge, ReachabilityLattice(Reachability.UNREACHABLE))
     }
+
+    // For all other edges, we simply propagate the reachability property of the edge which
+    // made us come here.
+    remainingEdges.forEach { state.push(it, state[enteringEdge]) }
+}
+
+/**
+ * Implements the [LatticeElement] over reachability properties: TOP | REACHABLE | UNREACHABLE |
+ * BOTTOM
+ */
+class ReachabilityLattice(override val elements: Reachability) :
+    LatticeElement<Reachability>(elements) {
+    override fun lub(other: LatticeElement<Reachability>) =
+        ReachabilityLattice(maxOf(this.elements, other.elements))
+
+    override fun duplicate() = ReachabilityLattice(this.elements)
+
+    override fun compareTo(other: LatticeElement<Reachability>) =
+        this.elements.compareTo(other.elements)
 }
+
+/** The ordering will be as follows: BOTTOM (no information) < UNREACHABLE < REACHABLE < TOP */
```

Comment:
Can you give information on what Top stands for like you did for Bottom?

Reply:
No because I did not even use it, so it didn't make sense. I just deleted it